### PR TITLE
Download in chunks concurrently.

### PR DESCRIPTION
**Pull request by psteiger**
Implements the feature of concurrent downloads of chunks when downloading a file, providing much better performance.

The implementation uses threads, coroutines (async/await) and asyncio library to achieve the goal.

Also, other minor changes.

Please help test drive it!

### CODE REVIEW

File: chunks.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        yield from self.chunks
+
+    def __getitem__(self, item):
+        return self.chunks[item]
+
+    @staticmethod
+    def create_from(file_downloader, file_name, file_size, chunk_size) -> 'Chunks':
+        number_of_chunks = calculate_number_of_chunks(file_size, chunk_size)
+        chunk_range = range(number_of_chunks)
+        # print(number_of_chunks, list(chunk_range))
```
Comment (viniciusalbuquerque) [inline chunks.py:67]: You can either use the `logger.d` or remove this line

File: command.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+            '--extract',
+            help='Tries to extract the downloaded file',
+            action='store_true'
+        )
+
+    async def execute(self):
+        extract = self.args.extract
+        if self.args.last:
+            await self.download_last_uploaded_file(extract)
+        elif not self.args.file:
```
Comment (viniciusalbuquerque) [inline command.py:84]: This is not a comment that is part of you change, but just to track it:

I think we could show the help for download here. I mean, if the user does `gdrive download` with no args, I think we could show the `gdrive download --help`

File: command.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+            await self.download_id(extract)
+        else:
+            await self.try_download_id_then_name(extract)
+
+    async def download_last_uploaded_file(self, extract):
+        last_file = self.google.get_last_modified_file()
+        if not last_file:
+            print("Could not find any file")
+            return
+        await self.download_from_metadata(last_file, extract)
```
Comment (viniciusalbuquerque) [inline command.py:99]: Should this be `return await self.download_from_metadata(last_file, extract)`? Actually, I believe that all of the others may not need to return.

File: command.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        return await self.download_filename(extract)
+
+    async def download_from_metadata(self, metadata, extract):
+        if metadata is None:
+            print('Could not find file to download')
+            return
+        file_name = metadata["name"] or "Unknown filename"
+        describe_files(metadata)
+        downloaded_file = await self.download(metadata)
+        if not downloaded_file:
```
Comment (viniciusalbuquerque) [inline command.py:128]: Is the extractor still working? Your download function isn't returning anything, it seems. That means that this
`if not download_file:` would be true always, so the code below would unreachable.

File: command.py (hunk 2 above)
Comment (psteiger) [reply command.py:84]: Agreed

File: chunks.py (hunk 1 above)
Comment (psteiger) [reply chunks.py:67]: Removed.

File: gdrive.py [hunk 5]
```diff
@@ -1,571 +1,18 @@
...
-        return None
-
-
-def main():
-    CommandParser().execute_command()
 
 if __name__ == '__main__':
-    main()
+    try:
+        asyncio.run(main())
```
Comment (jrcamelo) [inline gdrive.py:16]:
```
  File "./gdrive", line 16, in <module>
    asyncio.run(main())
AttributeError: module 'asyncio' has no attribute 'run'
```
Had to update to python 3.7 to be able to run it  
There are alternatives that supposedly work the same, such as  
`asyncio.get_event_loop().run_until_complete(main())`

File: chunks.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                for chunk in chunks:
+                    await chunk
+
+            def write_final_file_from_partial_files():
+                with open(file_name, "wb") as final_file:
+                    for chunk in chunks:
+                        with open(chunk.file_name, "rb") as partial_file:
+                            final_file.write(partial_file.read())
+
+            def remove_partial_files():
```
Comment (jrcamelo) [inline chunks.py:31]: I tested interrupting a download and it left part 31 here somehow  
It threw a lot of errors, and this one was at the end   
During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "/home/jrsc/Documents/gdrive-manager/chunks.py", line 39, in write_files
    remove_partial_files()
  File "/home/jrsc/Documents/gdrive-manager/chunks.py", line 33, in remove_partial_files
    os.remove(chunk.file_name)
FileNotFoundError: [Errno 2] No such file or directory: 'fastboot_sofia_vzw_userdebug_11_RPM31.107_eng.jcamelo.210415.165840_intcfg_test-keys_verizon_US.tar.gz.29'
```

File: command.py (hunk 4 above)
Comment (psteiger) [reply command.py:128]: Fixed

File: command.py (hunk 3 above)
Comment (psteiger) [reply command.py:99]: Instead, decided to return nothing, but raise exception. Fixed.

File: gdrive.py (hunk 5 above)
Comment (psteiger) [reply gdrive.py:16]:
```python
asyncio.get_event_loop().run_until_complete(main())
```

is not exactly the same: basically we would need to copy-paste the implementation of `run()` (which uses `run_until_complete`, but handles exceptions, shutdown, etc)

@jrcamelo what do you think about requiring Python 3.9? There are other APIs I intend on using... (`asyncio.to_thread`). It was introduced in 3.9.

File: chunks.py (hunk 6 above)
Comment (psteiger) [reply chunks.py:31]: Were all other parts removed?

Comment (jrcamelo) [reply chunks.py:31]: Yes, only part 31 left

File: config.py [hunk 7]
```diff
@@ -5,4 +5,5 @@
 CREDENTIALS_PATH = join(GDRIVE_PATH, 'credentials.json')
 TOKEN_PATH = join(GDRIVE_PATH, 'token.pickle')
 EXTRACTOR_CONFIG_FILE = join(GDRIVE_PATH, 'data_config.json')
-
+UPLOAD_CHUNK_SIZE = 1024 * 1024 * 10  # 10MB
+DOWNLOAD_CHUNK_SIZE = 1024 * 1024 * 10  # 10MB
```
Comment (carlosolimpio) [inline config.py:9]: What are your thoughts about these values for different internet speeds? And why did you pick these values?

Comment (psteiger) [reply config.py:9]: IQQ. Please run some tests with your setup. We might benefit from some logic to infer the chunk size optimally.

File: chunks.py (hunk 6 above)
Comment (psteiger) [reply chunks.py:36]: Fixed retro-compatibility with 3.6. Please retest.

Comment (psteiger) [reply chunks.py:36]: Fixed, please recheck and resolve if all is ok.

File: gdrive.py (hunk 5 above)
Comment (psteiger) [reply gdrive.py]: Fixed retro-compatibility with 3.6. Please retest.

**Review by viniciusalbuquerque — APPROVED** [review]

**Review by jrcamelo — APPROVED** [review]

**Pull request merged**